File: 01-array/0104-209.go
func minSubArrayLen(target int, nums []int) int {
    i := 0
    l := len(nums)  // size of array nums
    sum := 0        // sum
    res := l + 1 // the goal is to make a choice when no subarray is found
    for j := 0; j < l; j++ {
        sum += nums[j]
        for sum >= target {
            width := j - i + 1
            if width < res {
                res = width
            }
            sum -= nums[i]
            i++
        }
    }
    if res == l+1 {
        return 0
    } else {
        return res
    }
}

File: 01-array/0104-209.py
"""
leetcode 209

the sliding window is [left, right]
the width of sliding window is right-left+1

n = len(nums)
because the answer won't be greater than n+1, use it as a condition in case
that no subarray is found to satisfy the target

"""

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        res = n+1
        left = sum = 0
        for right, x in enumerate(nums):
            s += x
            while s >= target:
                res = min(res, right-left+1)
                s -= nums[left]
                left += 1
        return res if res <= n else 0
